### ISSUE TITLE
Line features with same text-field but different icon-image should not be merged

### ISSUE THREAD

**Issue by 1ec5** [mapbox/mapbox-gl-js#3645]
**mapbox-gl-js version**: 0.27.0

When the map is overzoomed, line features with identical `text-field` values but different `icon-image` values are incorrectly merged together.

For [this OpenStreetMap diary entry](http://www.openstreetmap.org/user/Minh%20Nguyen/diary/39123), I created a style based on the Mapbox Streets source that shows a different state route shield for the U.S. states of Ohio, Kentucky, and Indiana. The shield image is selected by assigning an `icon-image` of `us-state-{reflen}-{iso_3166_2}`.

<img width="741" alt="interstate" src="https://cloud.githubusercontent.com/assets/1231218/20381352/f4ea5292-ac5b-11e6-9f20-62e44397eff6.png">

In this screenshot, everything west of the state line should have the rectangular Indiana shield, while everything to the east of the state line should have the Ohio-shaped Ohio shield. Unfortunately, at zoom levels 16 and above – that is, beyond the Mapbox Streets source’s maximum zoom level – GL JS shows the Ohio shield on part of SR 252 on the Indiana side.

In OpenStreetMap, SR 252 is actually two ways: [210557386](http://www.openstreetmap.org/way/210557386) comes near the state line but doesn’t cross it, while the much shorter [203231052](http://www.openstreetmap.org/way/203231052) is a bridge that crosses the state line. Because the bridge’s center lies within Ohio, the entirety of 203231052 is tagged `iso_3166_2=US-OH` in the Mapbox Streets source. But 210557386 should always show the Indiana shield, because it’s tagged `iso_3166_2=US-IN` in the source.

@ajashton helped me track down the issue to mapbox/mapbox-gl-js#683, in which label lines are merged based on their text. In this case, both lines are apparently merged because they have the same text (`252`). However, each would have a different icon image, so they shouldn’t have been merged.

You can see the problem more clearly in this reduced test case: mapbox/mapbox-gl-test-suite@fdc08ed94f364a4fc76718aa12fbe1c13947d5bb. In this test case, I set `icon-image` to `bogus-{reflen}-{iso_3166_2}-18` and `text-field` to `{ref}`, similar to the style above. (I also set `text-opacity` to 0 for effect.) We should see the S’s give way to U’s as we go east to west:

![expected](https://cloud.githubusercontent.com/assets/1231218/20381231/609a0a1a-ac5b-11e6-9d76-9e88964faddc.png)

However, I actually see S’s all the way across:

![actual](https://cloud.githubusercontent.com/assets/1231218/20381234/62c65348-ac5b-11e6-90b0-847f7840be5a.png)

In fact, I generated the “expected” screenshot above by removing the `text-field` property. In the absence of `text-field`, either no merging occurs or the merging algorithm is more thorough somehow.

My use case requires a custom style, but it’s also the main use case for ISO 3166-2 codes on roads in the Mapbox Streets source. It’s also entirely possible that there are cases where this issue reproduces in an unmodified Mapbox Streets style. Suppose a U.S. route 25 has been decommissioned up to a state line, reverting to State Route 25 on one side, per standard practice. This would mean two adjoining lines would both have a `text-field` of `25`, but one would have a `shield` of `us-highway` while the other has `us-state`.

This issue also reproduces in gl-native master as of mapbox/mapbox-gl-native@f7db6cf7601e5d97492622cab76ed38ce692b37c.

/cc @mourner @jfirebaugh

**Comment by jfirebaugh** [comment]
Agreed, and once symbol layer acquires data-driven properties, this should check equality across all feature values seen in data-driven properties.

**Comment by ChrisLoer** [comment]
So... we want to replace `text` here:

https://github.com/mapbox/mapbox-gl-js/blob/180669e0ddf184a4c21363c1ecdadc8c2f81d803/src/symbol/mergelines.js#L37-L40

...with something like a hash of all `layout_symbol` properties for the feature (including `text-field` and `icon-image`), evaluated at the bucket's  tile zoom? We just started using murmurhash, so we have an easy way to collapse and store the result.

Is there a more minimal set we could use, or a way to avoid all the duplicate evaluation? Seems like a lot of extra work, although it might not turn out to be a big deal. Extracting the set of inputs to all `layout_symbol` properties and then just testing input equality might be faster, but also seems more complicated?

**Comment by 1ec5** [comment]
This overly aggressive merging might be the root cause of the glitch found in https://github.com/ZeLonewolf/openstreetmap-americana/pull/72#pullrequestreview-845993462 (which is based on MapLibre GL JS, but still): a `text-field` was set to a `formatted` expression that contained images separated by spaces, but no other text. The labels seemed to be pushed very far down the road from where they were expected, even beyond tile boundaries. Removing the spaces eliminated the issue. Maybe theembedded images were not considered for the purpose of conflating the LineString features, but removing the spaces caused the “plain text” representation to be the empty string, which disabled line merging.

/cc @jleedev
